ADBench/plot_graphs.py

```diff
@@ -192,176 +214,201 @@ def div_lists(alist, blist):
+
+    if do_save or do_plotly:
+        print(f"Output directory is: {out_dir}\n")
+
+        
+
+    rcParams.update({"figure.max_open_warning": 0})
+    figure_size = (9, 6) if do_plotly else (12, 8)
+
+    # Scan folder for all files, and determine which graphs to create
+    all_files = [path for path in utils._scandir_rec(in_dir) if TIMES_SUBSTRING in path[-1]]
+    all_graphs = [path.split("/") for path in set(["/".join(path[:-2]) for path in all_files])]
+    all_tool_names = tool_names(all_files)
+
+    function_types = ["objective ÷ Manual", "objective", "jacobian", "jacobian ÷ objective"]
+    all_graphs = [(path, function_type) for function_type in function_types for path in all_graphs]
+    all_graph_dict = {}
+
+
+
+    # Loop through each of graphs to be created
+    for (figure_idx, (graph, function_type)) in enumerate(all_graphs, start=1):
+        build_type = graph[0]
+        objective = graph[1]
+        maybe_test_size = graph[2:]
+
+        (graph_name, graph_save_location) = graph_data(build_type, objective, maybe_test_size)
+
+        # Create figure
+        figure = pyplot.figure(figure_idx, figsize=figure_size, dpi=fig_dpi)
+
+        # Extract file details
+        graph_files = [path for path in all_files if path[:len(graph)] == graph]
+
+        def sorting_key_fun(v):
+            y_list = utils.get_non_infinite_y_list(v[2])
+            if len(y_list) > 0:
+                return sum(y_list) / len(y_list)
+            else:
+                return 1e9
+
+        sorted_vals_by_tool = sorted(vals_by_tool(objective, graph_files, all_tool_names),
+                                    key=sorting_key_fun,
+                                    reverse=True)
+
+        handles, labels = [], []
+        violation_x, violation_y = [], []
+        was_violation = False
+        additional = []
+
+        # Plot results
+        for tool, n_vals, t_vals, violations in sorted_vals_by_tool:
+            def vals_with_neighbours_and_violation():
+                # Checking neighbours by shifting t_vals requires that
+                # it is in the order of monotonic n_vals
+                assert n_vals == sorted(n_vals)
+                return zip(
+                    n_vals,
+                    t_vals,
+                    # Whether the left neighbour is missing
+                    [True] + [t_val == float("inf") for t_val in t_vals],
+                    # Whether the right neighbour is missing
+                    [t_val == float("inf") for t_val in t_vals[1:]] + [True],
+                    violations)
+
+            color, marker = get_style(tool, all_tool_names)
+
+            label = utils.format_tool(tool)
+            all_terminated = all(t_val == float("inf") for t_val in t_vals)
+            # Append label in legend if all point values are infinite
+            if all_terminated:
+                label += ALL_TERMINATED_SUFFIX
+
+            labels.append(label)
+            handles += pyplot.plot(
```

Comment:
I think, it should be enough to not append the label if `t_vals` is empty.

Reply:
I suggest that we drop the change to avoid recalculating tool names.  Sure it's a little bit faster, but my main concern with this file is how easy it is to read.  It's already far more complicated than necessary and I think we should avoid changing the logic before we find a way of making it considerably simpler.

Reply:
> I suggest that we drop the change to avoid recalculating tool names. Sure it's a little bit faster, but my main concern with this file is how easy it is to read. It's already far more complicated than necessary and I think we should avoid changing the logic before we find a way of making it considerably simpler.

What did you mean by `avoid recalculating tool names`? Now they are calculated only once at the script beginning